earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +190,168 @@ class khMTSet
+  }
+  void erase(const K &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.erase(val);
+  }
+  size_t size(void) {
+    return map.size();
+  }
+  V& operator[](const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map[key];
+  }
+  void clear(void) {
+    std::lock_guard<std::mutex> guard(mutex);
+    map.clear();
+  }
+  std::vector<K> keys(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    std::vector<K> thekeys;
+    thekeys.reserve(map.size());
+    for (const auto& p : map) {
+      thekeys.push_back(p.first);
+    }
+    return thekeys;
+  }
+  bool contains(const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.find(key) != map.end();
+  }
+  bool empty(void) const {
+    return map.empty();
+  }
 };
 
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but creates some less obvious ones
+template <typename T>
+class MTVector : public std::vector<T>{
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
+
+  private:
+    mutable khRWLock mtx;
+
+    MTVector(const MTVector& a, const khReadGuard&) : Base(a) { }
+
+  public:
+    void push_back(const T& v) {
+      khWriteGuard lock(mtx);
+      Base::push_back( v );
+    }
+    MTVector() { }
+    // forward to private copy constructor to protect vector from modification
+    MTVector(const MTVector& a) : MTVector(a, khReadGuard(a.mtx)) { }
+    MTVector(const std::vector<T>& v) : Base(v) { }
+    
+
+    const T& operator[] (const size_t idx) const {
+      khReadGuard lock(mtx);
+      return Base::operator[] (idx); 
+    }
+
+    bool operator==(const MTVector& x) const {
+      // Since using a read lock, no need to worry about lock ordering or things like a==a
+      khReadGuard lock1(mtx);
+      khReadGuard lock2(x.mtx);
+      return std::equal(Base::begin(), Base::end(), x.begin());
+    }
+
+    bool operator==(const Base& x) const {
+      khReadGuard lock(mtx);
+      return Base::operator==(x);
+    }
+
+    void clear(void) {
+      khWriteGuard lock(mtx);
+      Base::clear();
+    }
+
+    size_t size(void) const {
+      khWriteGuard lock(mtx);
+      return Base::size();
+    }
+
+    bool empty(void) const {
+      khWriteGuard lock(mtx);
+      return Base::empty();
+    }
+
+    void shrink_to_fit(void) {
+      khWriteGuard lock(mtx);
+      Base::shrink_to_fit();
+    }
+    
+    
+    void doForEach(std::function<void (const T&)> func) const {
+      if (!func) return; // can't do anyting with an invalid function
+      khReadGuard lock(mtx);
+      for_each(Base::begin(), Base::end(), func);
+    }
+
+    bool doForEachUntil(std::function<bool (const T&)> func) const {
+      if (!func) return false; // can't do anyting with an invalid function
+      khReadGuard lock(mtx);
+      bool loopedThroughAll = true;
+      for (uint i = 0; i < Base::size(); ++i) {
+        if (!func(operator [] (i))) {
```

Comment:
I'm not clear on the need to iterate via indices instead of an iterator. Also, calling `operator[]` here is going to attempt to apply another read lock. It should work, but that's a little bit of additional overhead for each access.